2022Q4 Wireless report (#559)
freebsd/freebsd-quarterly: 2022q4/wifi.adoc

=== Wireless updates

Links: +
link:https://people.freebsd.org/~bz/wireless/[Bjoern's Wireless Work In Progress landing page] URL: link:https://people.freebsd.org/\~bz/wireless/[https://people.freebsd.org/~bz/wireless/] +

Contact: Bjoern A. Zeeb <bz@FreeBSD.org>

> During the 4th quarter of 2022 not much work was publicly visible and admittedly slightly slow.

Reviewer: > During the 4th quarter of 2022 

Instead: 

> During the quarter

Author: That makes no sense;  which of the 4 quarters?  During the last quarter of the year?

Reviewer: It's a minor suggestion, for conciseness. 

> … During the last quarter of the year?

The period will be given in the introduction to the report, there's not yet an `intro.adoc` at <https://github.com/freebsd/freebsd-quarterly/tree/main/2022q4>. 

Produced from <https://github.com/freebsd/freebsd-quarterly/blob/main/2022q3/intro.adoc#readme>, a shot of the head of the report for the third quarter: 

![image](https://user-images.githubusercontent.com/192271/211212008-d461eb5f-dc3a-4539-b623-e0abdb0eecb1.png)

There's context – the period – in three places at beginning: 

1. the title bar (or tab)
2. the first heading
3. the first sentence

– and the title bar (or tab) context should remain visible throughout.

Reviewer: Resolved by [`12e0e61` (#559)](https://github.com/freebsd/freebsd-quarterly/pull/559/commits/12e0e61a7af41468c093a6ae1fd82ee77fed7b48), thanks.

> During the 4th quarter of 2022 not much work was publicly visible and admittingly slightly slow.

Reviewer 2: s/admittingly/admittedly/

Behind the scenes wireless work was happening on two fronts:

>  * work on 11n and 11ac and wpa bits,

Reviewer: >  * work on 11n and 11ac and wpa bits,

Instead: 

> * 11n and 11ac and wpa

>  * and continous work on more drivers and firmware, problems, testing, and filling more gaps for these.

Reviewer: >  * and continous work on more drivers and firmware, problems, testing, and filling more gaps for these.

Instead: 

> * more drivers and firmware, problems, testing, and filling gaps for these.

(The word _work_ is in the sentence that preceded these two bullet points.)

While the main development for newer standards and the Intel iwlwifi driver is sponsored by the FreeBSD Foundation, I find myself spending a lot of private time with Realtek rtw88 and rtw89, Atheros, and Mediatek mt76 (7921 and 7915) drivers now as well.
Testing changes on bare metal and using bhyve VMs using passthru has become more time-consuming, with the amount of supported chipsets increasing, in order not to break other drivers.

> Even different (generations of) chipsets supported by the same driver, at times, behave different to the same change.

Reviewer 2: s/different to/differently with/ (or s/behave different/react differently/)

A separate discussion was brought to me about the size of firmware added to the tree for the already existing drivers and the firmware for more drivers coming.

> The chicken-egg problem to solve is having firmware available on the release media; without firmware, a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration.

Reviewer: @bzfbd please, do you mean that the problem began (in the second quarter) when `13.1-RELEASE` was released; that (without the `RELEASE` medium context) there's no problem with the size of firmware added to the tree?

I imagine not; this is part of what's not clear.

> The chicken-egg problem to solve is to have the firmware available on the release media as otherwise a lot of modern laptops will not have any sort of outside communications available at the time of install in their default ocnfiguration should we not commit it to main anymore.

Reviewer 2: s/ocnfiguration/configuration/

> The chicken-egg problem to solve is to have the firmware available on the release media as otherwise a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration should we not commit it to main anymore.

Reviewer: > The chicken-egg problem to solve is to have the firmware available on the release media as otherwise a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration should we not commit it to main anymore.

That's a relatively long sentence, with no commas. 

Also, it's ambiguous; as if including firmware _is_ the problem.

Reviewer 2: I don't see ambiguity here, but that may be only me.
```suggestion
The chicken-egg problem to solve is to have the firmware available on the release media; without it, a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration.
```

Reviewer: A _solution_ to a problem will be to do something. In this case, what's done is: 

* having firmware.

I see: "the _problem_ is to have firmware".

Reviewer 2: Which is correct, because of the limited space on distribution media. @bzfbd, is that what you meant?

This will be a larger discussion to have to also solve firmware for other drivers, but that discussion will be for another day and place.

> Slightly belated I have started to flush LinuxKPI and 802.11 changes into the tree and of the year and that work will drag into early 2023 at which point more of the aforementioned remaining drivers will also hit the tree.

Reviewer 2: s/belated/belatedly/

s/flush/push/ maybe?

s/and of the year/at the end of the year/

> Slightly belatedly I have started to push LinuxKPI and 802.11 changes into the tree at the end of the year and that work will drag into early 2023 at which point more of the aforementioned remaining drivers will also hit the tree.

Reviewer: > drag

Instead: 

> continue

Reviewer 2: I'm fine with "drag" if the connotation of (mild) annoyance at the time it's taking is intended, otherwise "continue" is better.

One of the main remaining problems to solve is the firmware crashes on interface down/up cycles currently experienced with at least two drivers.

Thankfully during the last weeks, after my call for help, multiple people have stood up wanting to help with various drivers (especially Realtek and Mediatek).
I hope that after me catching up and pushing things out this can accelerate progress again.

Thanks again to everyone doing testing, providing debug output, sending in feedback, or using the drivers at this point.

> For the latest state of the development, please follow updates posted to the freebsd-wireless mailing list and check the landing page which has links to all wiki pages for each driver status.

Reviewer: > the landing page which has

Instead: 

> the landing page, which has

Reviewer: > follow updates posted to the freebsd-wireless mailing list and check

Instead: 

> use the freebsd-wireless mailing list, and check

Contact: Bjoern A. Zeeb <bz@FreeBSD.org>

During the quarter not much work was publicly visible and admittedly slightly slow.
Behind the scenes wireless work was happening on two fronts:

 * 11n, 11ac, and wpa,

 * more drivers and firmware, problems, testing, and filling gaps for these.

While the main development for newer standards and the Intel iwlwifi driver is sponsored by the FreeBSD Foundation, I find myself spending a lot of private time with Realtek rtw88 and rtw89, Atheros, and Mediatek mt76 (7921 and 7915) drivers now as well.
Testing changes on bare metal and using bhyve VMs using passthru has become more time-consuming, with the amount of supported chipsets increasing, in order not to break other drivers.
Even different (generations of) chipsets supported by the same driver, at times, behave differently with the same change.

A separate discussion was brought to me about the size of firmware added to the tree for the already existing drivers and the firmware for more drivers coming.

> The chicken-egg problem to solve is having firmware available on the release media; without a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration.

Reviewer 2: ```suggestion
The chicken-egg problem to solve is having firmware available on the release media; without firmware, a lot of modern laptops will not have any sort of outside communications available at the time of install in their default configuration.
```
